### PR TITLE
ci: mitigate flapping block device in ceph test suites

### LINKED ISSUE #7405: ci: smoke test times out on listing rbd images
Seeing in multiples job like this one: https://github.com/rook/rook/pull/7103/checks?check_run_id=2096387561

Reported also here: https://github.com/rook/rook/pull/7400#issuecomment-799085217

Logs:

```
2021-03-15T01:49:44.7896081Z === RUN TestCephSmokeSuite/TestBlockStorage_SmokeTest
2021-03-15T01:49:44.7905141Z 2021-03-15 01:49:44.789998 I | integrationTest: Block Storage End to End Integration Test - create, mount, write to, read from, and unmount
2021-03-15T01:49:44.7906815Z 2021-03-15 01:49:44.790015 I | integrationTest: Running on Rook Cluster smoke-ns
2021-03-15T01:49:44.7907806Z 2021-03-15 01:49:44.790019 I | integrationTest: Step 0 : Get Initial List Block
2021-03-15T01:49:44.7909156Z 2021-03-15 01:49:44.790028 D | exec: Running command: kubectl exec -i rook-ceph-tools -n smoke-ns -- ceph osd lspools --connect-timeout=15 --format json
2021-03-15T01:49:46.0122492Z 2021-03-15 01:49:46.011781 D | exec: Running command: kubectl exec -i rook-ceph-tools -n smoke-ns -- rbd ls -l device_health_metrics --format json
2021-03-15T02:17:24.9291728Z panic: test timed out after 30m0s
```

When logging into the runner I could see that the OSD was crash looping for the following reason:

```
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  0 ceph version 15.2.9 (357616cbf726abb779ca75a551e8d02568e15b17) octopus (stable), process ceph-osd, pid 1
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  0 pidfile_write: ignore empty --pid-file
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) init /var/run/ceph/ceph-osd.0.asok
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) bind_and_listen /var/run/ceph/ceph-osd.0.asok
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) register_command 0 hook 0x5595e8ee1228
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) register_command version hook 0x5595e8ee1228
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) register_command git_version hook 0x5595e8ee1228
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) register_command help hook 0x5595e8edc4c0
debug 2021-03-12T16:18:25.135+0000 7ff219bdaf40  5 asok(0x5595e8f90000) register_command get_command_descriptions hook 0x5595e8edc070
debug 2021-03-12T16:18:25.139+0000 7ff219bdaf40  5 object store type is bluestore
debug 2021-03-12T16:18:25.139+0000 7ff219bdaf40 10 bluestore(/var/lib/ceph/osd/ceph-0) set_cache_shards 1
debug 2021-03-12T16:18:25.139+0000 7ff219bdaf40 10 bluestore(/var/lib/ceph/osd/ceph-0/block) _read_bdev_label
debug 2021-03-12T16:18:25.139+0000 7ff219bdaf40 -1 bluestore(/var/lib/ceph/osd/ceph-0/block) _read_bdev_label failed to open /var/lib/ceph/osd/ceph-0/block: (13) Permission denied
debug 2021-03-12T16:18:25.139+0000 7ff219bdaf40 -1  ** ERROR: unable to open OSD superblock on /var/lib/ceph/osd/ceph-0: (2) No such file or directory
debug 2021-03-12T16:18:25.139+0000 7ff212146700  5 asok(0x5595e8f90000) entry start
debug 2021-03-12T16:18:25.139+0000 7ff212146700 20 asok(0x5595e8f90000) entry waiting
```

However the permission seems correct:

```
runner@fv-az183-677:~/work/rook/rook$ ls /var/lib/rook/smoke-ns/_dev_sda1/ -al
total 32
drwxr-xr-x 2  167  167 4096 Mar 12 16:02 .
drwxr-xr-x 3 root root 4096 Mar 12 16:02 ..
lrwxrwxrwx 1  167  167    9 Mar 12 16:02 block -> /dev/sda1
-rw------- 1  167  167   37 Mar 12 16:02 ceph_fsid
-rw------- 1  167  167   37 Mar 12 16:02 fsid
-rw------- 1  167  167   55 Mar 12 16:02 keyring
-rw------- 1  167  167    6 Mar 12 16:02 ready
-rw------- 1  167  167   10 Mar 12 16:02 type
-rw------- 1  167  167    2 Mar 12 16:02 whoami
```

Need to dive a bit more, for whoever is interested in the debugging the runner directly, you can add:

```
    - name: Setup tmate session
      uses: mxschmitt/action-tmate@v3
```

In the test file before the `go test` runs.
Then look at the tmate logs, op into the VM and run `touch continue`, the test will run, then wait and explore the env while it is running.

@satoru-takeuchi FYI.

### PR DESCRIPTION
**Pull request by satoru-takeuchi**
**Description of your changes:**

`rbd ls -l device_health_metrics` causes timeout frequently.

```console
...
2021-03-15T01:49:44.7896081Z === RUN TestCephSmokeSuite/TestBlockStorage_SmokeTest
2021-03-15T01:49:44.7905141Z 2021-03-15 01:49:44.789998 I | integrationTest: Block Storage End to End Integration Test - create, mount, write to, read from, and unmount
2021-03-15T01:49:44.7906815Z 2021-03-15 01:49:44.790015 I | integrationTest: Running on Rook Cluster smoke-ns
2021-03-15T01:49:44.7907806Z 2021-03-15 01:49:44.790019 I | integrationTest: Step 0 : Get Initial List Block
2021-03-15T01:49:44.7909156Z 2021-03-15 01:49:44.790028 D | exec: Running command: kubectl exec -i rook-ceph-tools -n smoke-ns -- ceph osd lspools --connect-timeout=15 --format json
2021-03-15T01:49:46.0122492Z 2021-03-15 01:49:46.011781 D | exec: Running command: kubectl exec -i rook-ceph-tools -n smoke-ns -- rbd ls -l device_health_metrics --format json
2021-03-15T02:17:24.9291728Z panic: test timed out after 30m0s
...
```

It's because the block device file, that backs OSD, is attached/detached continuously in some github actions runners. This problem seem to happen if there is no mounted filesystem in this device. Let's bypass this problem by mounting a fake filesystem.

**Which issue is resolved by this Pull Request:**
Resolves #7405 

**Checklist:**

- [o] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [o] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [o] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [o] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [o] Documentation has been updated, if necessary.
- [o] Unit tests have been added, if necessary.
- [o] Integration tests have been added, if necessary.
- [o] Pending release notes updated with breaking and/or notable changes, if necessary.
- [o] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [o] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by leseb — COMMENTED** [review]
FYI: https://github.com/rook/rook/issues/7405

**Comment by satoru-takeuchi** [comment]
@leseb I'll investigate #7405 , if it would be able to fix the root cause. I mark.thos PR as draft for now.

File: tests/framework/clients/block.go [hunk 1]
```diff
@@ -151,6 +151,11 @@ func (b *BlockOperation) ListAllImages(clusterInfo *client.ClusterInfo) ([]Block
 	// for each pool, get further details about all the images in the pool
 	images := []BlockImage{}
 	for _, p := range pools {
+		// Let's ignore to list device_health_metrics pool that sometimes waits forever.
+		// This filtering is harmless because this pool is empty in new cluster.
+		if p.Name == "device_health_metrics" {
+			break
```
Comment (travisn) [inline tests/framework/clients/block.go:157]: I think we want to continue with other pools and just ignore this one, right?
```suggestion
			continue
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by travisn** [comment]
@satoru-takeuchi Any more clues on this one? It's our most common test issue now, would be great to fix...

**Comment by satoru-takeuchi** [comment]
@travisn 

I heard @leseb will fix this problem soon. CMIIW.

**Comment by leseb** [comment]
> @travisn
> 
> I heard @leseb will fix this problem soon. CMIIW.
> 
> If it would take more times, it's also OK to bypass this problem with this PR for now.

I got distracted by other things. I might only have time next week to come back to this.
I'm not sure if this patch actually bypasses the issue. I believe it only delays the problem until a pool is created.

The problem is that the OSD is crash looping because the permissions on the disk are reset (by udev most likely). So the cluster is never in a good shape. The main problem is that not all runners have this issue.

**Comment by satoru-takeuchi** [comment]
@leseb Yes,I confirmed that this problem doesn't disappear by this PR. And now I'm working on this issue. Hope it will be fixed in this week. This issue seems to happen frequently very much today.

**Comment by leseb** [comment]
> @leseb Yes,I confirmed that this problem doesn't disappear by this PR. And now I'm working on this issue. Hope it will be fixed in this week. This issue seems to happen frequently very much today.

Thanks for taking over.

**Comment by satoru-takeuchi** [comment]
Tmate session after building rook.

```console
$ ls -l /dev/sd*
brw-rw---- 1 root disk 8,  0 Mar 24 09:55 /dev/sda
brw-rw---- 1 root disk 8,  1 Mar 24 09:55 /dev/sda1
brw-rw---- 1 root disk 8, 14 Mar 24 09:55 /dev/sda14
brw-rw---- 1 root disk 8, 15 Mar 24 09:55 /dev/sda15
brw-rw---- 1 root disk 8, 16 Mar 24 10:05 /dev/sdb
brw-rw---- 1 root disk 8, 17 Mar 24 10:05 /dev/sdb1
```

```console
$ dmesg | grep sdb1
[    2.645945]  sdb: sdb1
[   10.131548]  sdb: sdb1
[   11.863341] EXT4-fs (sdb1): mounted filesystem with ordered data mode. Opts: (null)
```

**Comment by satoru-takeuchi** [comment]
@leseb I have basic questions about host-based cluster because I'm not familiar with this type cluster. I consider the expected owner:group of block devices for OSD is "root:disk" than "ceph(167):ceph(167)"?  If "ceph:ceph", is there any logic to override existing block device files with new ones?

**Comment by leseb** [comment]
> @leseb I have basic questions about host-based cluster because I'm not familiar with this type cluster. I consider the expected owner:group of block devices for OSD is "root:disk" than "ceph(167):ceph(167)"? If "ceph:ceph", is there any logic to override existing block device files with new ones?

The `root:disk` permission is the system's default. The reason why we override it to `ceph:ceph` (or `167:167`) is because the OSD daemon run under the `ceph` user and not root.
Does that help?

**Comment by satoru-takeuchi** [comment]
> The root:disk permission is the system's default. The reason why we override it to ceph:ceph (or 167:167) is because the OSD daemon run under the ceph user and not root.
> Does that help?

@leseb Ah, yes, thank you! I found the chown in osd code just now.

https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/spec.go#L103

I'll continue to investigate this issue especially focusing on udev.

**Comment by leseb** [comment]
> > The root:disk permission is the system's default. The reason why we override it to ceph:ceph (or 167:167) is because the OSD daemon run under the ceph user and not root.
> > Does that help?
> 
> @leseb Ah, yes, thank you! I found the chown in osd code just now.
> 
> https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/spec.go#L103
> 
> I'll continue to investigate this issue especially focusing on udev.

@satoru-takeuchi I just thought about something that might be a simple fix here: https://github.com/rook/rook/pull/7467, let's see how that work!

**Comment by satoru-takeuchi** [comment]
@leseb Wow, it's very simple and would be enough to workaround this problem. I'll run HelmSuite and SmokeSuite multiple times to confirm how it works.

**Comment by satoru-takeuchi** [comment]
@leseb When failure, sdb seems to be attached/detached continuously.

```console
dmesg
....
[ 1116.905811] docker0: port 4(veth40b3653) entered blocking state
[ 1116.905812] docker0: port 4(veth40b3653) entered forwarding state
[ 1119.483639]  sdb: sdb1
[ 1119.984324]  sdb: sdb1
[ 1124.583154]  sdb: sdb1
[ 1125.103205]  sdb: sdb1
[ 1129.676603]  sdb: sdb1
[ 1130.102950]  sdb: sdb1
[ 1134.776521]  sdb: sdb1
[ 1135.250131]  sdb: sdb1
[ 1139.854729]  sdb: sdb1
[ 1140.258897]  sdb: sdb1
[ 1144.942330]  sdb: sdb1
[ 1145.402839]  sdb: sdb1
[ 1150.033066]  sdb: sdb1
[ 1150.434360]  sdb: sdb1
[ 1155.138325]  sdb: sdb1
[ 1155.701090]  sdb: sdb1
[ 1160.217446]  sdb: sdb1
[ 1160.614816]  sdb: sdb1
[ 1165.293167]  sdb: sdb1
[ 1165.700072]  sdb: sdb1
[ 1170.359880]  sdb: sdb1
[ 1170.746992]  sdb: sdb1
[ 1175.430908]  sdb: sdb1
[ 1175.824292]  sdb: sdb1
[ 1180.525786]  sdb: sdb1
[ 1180.976678]  sdb: sdb1
[ 1185.607651]  sdb: sdb1
[ 1185.972740]  sdb: sdb1
[ 1190.719243]  sdb: sdb1
[ 1191.200250]  sdb: sdb1
[ 1195.794035]  sdb: sdb1
[ 1196.226549]  sdb: sdb1
[ 1200.875002]  sdb: sdb1
[ 1201.293201]  sdb: sdb1
[ 1205.953587]  sdb: sdb1
[ 1206.361987]  sdb: sdb1
[ 1211.024444]  sdb: sdb1
[ 1211.405058]  sdb: sdb1
[ 1216.121915]  sdb: sdb1
[ 1216.525097]  sdb: sdb1
[ 1221.203612]  sdb: sdb1
[ 1221.641304]  sdb: sdb1
```

Then block device files are re-created each time and all owner information are cleared. It would be a VM instance matter than udev, but I'm not sure yet. Hope your PR fix this problem.

**Comment by satoru-takeuchi** [comment]
I'll try to copy the target device to /dev/rook-test. It might be able to avoid continuous device file re-creation.

```
    - name: TestCephHelmSuite
      run: |
       tests/scripts/minikube.sh helm
       tests/scripts/helm.sh up
       DEV=$(lsblk|awk '/14G/ {print $1}'| head -1)
       cp /dev/${DEV} /dev/rook-test
       export DEVICE_FILTER=rook-test
       SKIP_TEST_CLEANUP=false SKIP_CLEANUP_POLICY=false go test -v -timeout 1800s -run CephHelmSuite github.com/rook/rook/tests/integration
```

**Comment by satoru-takeuchi** [comment]
> I'm trying to copy the target device to /dev/rook-test. It might be able to avoid continuous device file re-creation.

It's impossible since copied device files are not recognized by lsblk.

**Comment by satoru-takeuchi** [comment]
@leseb Finally I succeeded to bypass this problem. PTAL.

**Comment by satoru-takeuchi** [comment]
It will be resolved by #7467 in a different way.

**Pull request closed without merging**